p2p/discovery.py

```diff
@@ -234,9 +247,134 @@ def send_neighbours(self, node: kademlia.Node, neighbours: List[kademlia.Node])
+        super().__init__(privkey, address, bootstrap_nodes)
+
+        if preferred_nodes is not None:
+            self.preferred_nodes = preferred_nodes
+        else:
+            self.logger.debug('Operating with no preferred nodes')
+            self.preferred_nodes = tuple()
+
+        self._preferred_node_tracker = collections.defaultdict(lambda: 0)
+
+    @to_tuple
+    def _get_eligible_preferred_nodes(self) -> Iterator[kademlia.Node]:
+        """
+        Return nodes from the preferred_nodes which have not been used within
+        the last preferred_node_recycle_time
+        """
+        for node in self.preferred_nodes:
+            last_used = self._preferred_node_tracker[node]
+            if time.time() - last_used > self.preferred_node_recycle_time:
+                yield node
+
+    def _get_random_preferred_node(self) -> kademlia.Node:
+        """
+        Return a random node from the preferred list.
+        """
+        eligible_nodes = self._get_eligible_preferred_nodes()
+        if not eligible_nodes:
+            raise NoEligibleNodes("No eligible preferred nodes available")
+        node = random.choice(eligible_nodes)
+        return node
+
+    def get_random_bootnode(self) -> Iterator[kademlia.Node]:
+        """
+        Return a single node to bootstrap, preferring nodes from the preferred list.
+        """
+        try:
+            node = self._get_random_preferred_node()
+            self._preferred_node_tracker[node] = time.time()
+            yield node
+        except NoEligibleNodes:
+            yield from super().get_random_bootnode()
+
+    def get_nodes_to_connect(self, count: int) -> Iterator[kademlia.Node]:
+        """
+        Return up to `count` nodes, preferring nodes from the preferred list.
+        """
+        preferred_nodes = self._get_eligible_preferred_nodes()[:count]
+        for node in preferred_nodes:
+            self._preferred_node_tracker[node] = time.time()
+            yield node
+
+        num_nodes_needed = max(0, count - len(preferred_nodes))
+        yield from super().get_nodes_to_connect(num_nodes_needed)
+
+
+class DiscoveryService(BaseService):
+    logger = logging.getLogger("p2p.discovery.DiscoveryService")
+    _lookup_running = asyncio.Lock()
+    _last_lookup: float = 0
+    _lookup_interval: int = 30
+
+    def __init__(self, proto: DiscoveryProtocol, peer_pool: PeerPool) -> None:
+        super().__init__()
+        self.proto = proto
+        self.peer_pool = peer_pool
+
+    async def _run(self) -> None:
+        connect_loop_sleep = 2
+        asyncio.ensure_future(self.proto.bootstrap())
+        while True:
+            await self.maybe_connect_to_more_peers()
+            await self.wait(asyncio.sleep(connect_loop_sleep))
+
+    async def maybe_connect_to_more_peers(self) -> None:
+        """Connect to more peers if we're not yet maxed out to max_peers"""
+        num_connected_nodes = len(self.peer_pool)
```

Comment:
Can you move this inside the `if` statement (or maybe just inline with the logging statement)